diff --git a/lib/quick-sort.js b/lib/quick-sort.js
index 23f9eda57f7383b5291361e4593cf2fdef5409ba..a860b41503761d8543a4f68d6b0999d9e51a9dd3 100644
--- a/lib/quick-sort.js
+++ b/lib/quick-sort.js
@@ -107,8 +107,9 @@ function doQuickSort(ary, comparator, p, r) {
 }
 
 function cloneSort(comparator) {
-  let template = SortTemplate.toString();
-  let templateFn = new Function(`return ${template}`)();
+  // let template = SortTemplate.toString();
+  // let templateFn = new Function(`return ${template}`)();
+  let templateFn = (...args) => SortTemplate(...args);
   return templateFn(comparator);
 }
 